### PR TITLE
Filter maps produced by `GraphPlacement` and `NoiseAwarePlacement` to leave some `Qubit` unassigned

### PR DESCRIPTION
**Pull request by sjdilkes**
We've observed before that dynamically assign Program Qubits to Architecture Nodes during routing can give an improvement in performance. In this PR I've added an extra step to the end of `GraphPlacement::get_all_placement_maps`, where it removes some Qubit->Node assignments from the constructed maps under some basic condition.

### CODE REVIEW

File: tket/src/Placement/GraphPlacement.cpp [hunk 1]
```diff
@@ -283,6 +283,58 @@ GraphPlacement::get_all_weighted_subgraph_monomorphisms(
...
+        auto [_, exists] =
+            boost::edge(entry_vertex, *vertex_iter_begin, pattern_graph);
+        if (exists) {
+          n_interacting++;
+        }
+      }
+    }
+    if (std::min(n_pattern_edges, n_target_edges) - n_interacting <=
+        n_interacting) {
+      out_map.insert({entry.first, entry.second});
```
[the remark this answers is not in the document]
Comment (sjdilkes) [reply tket/src/Placement/GraphPlacement.cpp:332]: I think I'm OK with it: my concern is that if we update `bimap` that we will have a knock on effect where chains of `Qubit` get unassigned. If we need to dynamically assign "pairs" of qubits at a time then we have less useful information and so I suspect worse results. Basically I think it's _probably_ better if we have fewer unassigned.

[the remark this answers is not in the document]
Comment (yao-cqc) [reply tket/src/Placement/GraphPlacement.cpp:329]: Thanks. I agree it's probably better not to filter this map too aggressively.

**Comment by yao-cqc** [comment]
Looks like the test failure is caused by a compatibility issue between pytest 7.2.0 (released 4 hours ago) [it no longer depends on the py library](https://github.com/pytest-dev/pytest/releases ) and [pytest-benchmark]( https://github.com/ionelmc/pytest-benchmark/blob/073d7edbd92691e110d45c7a10e027cbf28a18ba/src/pytest_benchmark/logger.py#L20). I think we can pin pytest for now or install the `py` library separately. @cqc-alec what do you think?

**Comment by cqc-alec** [comment]
> Looks like the test failure is caused by a compatibility issue between pytest 7.2.0 (released 4 hours ago) [it no longer depends on the py library](https://github.com/pytest-dev/pytest/releases) and [pytest-benchmark](https://github.com/ionelmc/pytest-benchmark/blob/073d7edbd92691e110d45c7a10e027cbf28a18ba/src/pytest_benchmark/logger.py#L20). I think we can pin pytest for now or install the `py` library separately. @cqc-alec what do you think?

Maybe for now we can add `py` to the list in `pytket/tests/requirements.txt`.

**Review by yao-cqc — APPROVED** [review]

**Pull request merged**